crates/holochain/src/core/workflow/app_validation_workflow/unit_tests.rs

```diff
@@ -0,0 +1,506 @@
+    timestamp::Timestamp,
+    validate::ValidationStatus,
+    zome::{IntegrityZomeDef, Zome},
+    Action,
+};
+use kitsune_p2p_types::ok_fut;
+use matches::assert_matches;
+use parking_lot::{Mutex, RwLock};
+use std::{
+    collections::HashSet,
+    sync::{
+        atomic::{AtomicI8, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+// test app validation with a must get action
+// where initially the action is not in the cache db
+// and is then added to it
+#[tokio::test(flavor = "multi_thread")]
+async fn validation_callback_must_get_action() {
+    let zomes =
+        SweetInlineZomes::new(vec![], 0).integrity_function("validate", move |api, op: Op| {
+            if let Op::RegisterAgentActivity(RegisterAgentActivity { action, .. }) = op {
+                if let Ok(_) = api.must_get_action(MustGetActionInput::new(action.to_hash())) {
+                    Ok(ValidateCallbackResult::Valid)
+                } else {
+                    Ok(ValidateCallbackResult::UnresolvedDependencies(
+                        UnresolvedDependencies::Hashes(vec![action.to_hash().into()]),
+                    ))
+                }
+            } else {
+                Ok(ValidateCallbackResult::Invalid("wrong op type".to_string()))
+            }
+        });
+
+    let TestCase {
+        agent_key,
+        keystore,
+        integrity_zomes,
+        test_space,
+        ribosome,
+        workspace_read,
+        ..
+    } = TestCase::new(zomes).await;
+
+    let mut create = fixt!(Create);
+    create.author = agent_key.clone();
+    let create_action = Action::Create(create.clone());
+    let action_hashed = ActionHashed::from_content_sync(Action::Create(create));
+    let signed_action = SignedActionHashed::sign(&keystore, action_hashed)
+        .await
+        .unwrap();
+    let op = Op::RegisterAgentActivity(RegisterAgentActivity {
+        action: signed_action,
+        cached_entry: None,
+    });
+    let zomes_to_invoke = ZomesToInvoke::one_integrity(integrity_zomes[0].clone());
+    let invocation = ValidateInvocation::new(zomes_to_invoke, &op).unwrap();
+
+    let network = fixt!(HolochainP2pDna);
+    let fetched_dependencies = Arc::new(Mutex::new(HashSet::new()));
+
+    // action has not been written to a database yet
+    // validation should indicate it is awaiting create action hash
+    let outcome = run_validation_callback_inner(
+        invocation.clone(),
+        &ribosome,
+        workspace_read.clone(),
+        network.clone(),
+        fetched_dependencies.clone(),
+    )
+    .await
+    .unwrap();
+    assert_matches!(outcome, Outcome::AwaitingDeps(hashes) if hashes == vec![create_action.to_hash().into()]);
+
+    // write action to be must got during validation to dht cache db
+    let dht_op = DhtOp::RegisterAgentActivity(fixt!(Signature), create_action.clone());
+    let dht_op_hashed = DhtOpHashed::from_content_sync(dht_op);
+    test_space.space.cache_db.test_write(move |txn| {
```

Comment:
This will certainly work, but for correctness, data that we're validating should be directed to the DHT database right?

Reply:
This speaks to the weirdness of this test, but no this should indeed be the cache, because things we must_get go into the cache, and we don't validate them. It's just weird because this test needs to must_get the thing it's already validating, so it winds up in the cache and the dht db (after passing validation which can only happen after the action is in the cache.)